syntax = "proto3";
package proto;

import "google/protobuf/empty.proto";
import "google/protobuf/timestamp.proto";

message PostStatsRequest {
    int32 post_id = 1;
}

message StatsResponse {
    int64 views = 1;
    int64 likes = 2;
    int64 comments = 3;
}

message DynamicRequest {
    int32 post_id = 1;
}

message DayCount {
    google.protobuf.Timestamp day = 1;
    int64 count = 2;
}

message DynamicResponse {
    repeated DayCount items = 1;
}

message TopRequest {
    enum StatType {
        VIEWS = 0;
        LIKES = 1;
        COMMENTS = 2;
    }
    StatType type = 1;
}

message PostItem {
    int32 post_id = 1;
    int64 count = 2;
}

message UserItem {
    int32 user_id = 1;
    int64 count = 2;
}

message TopPostsResponse {
    repeated PostItem posts = 1;
}

message TopUsersResponse {
    repeated UserItem users = 1;
}

service StatsService {
    rpc GetPostStats(PostStatsRequest) returns (StatsResponse);
    rpc GetPostViewsDynamic(DynamicRequest) returns (DynamicResponse);
    rpc GetPostLikesDynamic(DynamicRequest) returns (DynamicResponse);
    rpc GetPostCommentsDynamic(DynamicRequest) returns (DynamicResponse);
    rpc GetTopPosts(TopRequest) returns (TopPostsResponse);
    rpc GetTopUsers(TopRequest) returns (TopUsersResponse);
}
